### ISSUE TITLE
iframe: various methods

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [arkenfox/TZP#17]
what are all these methods and access? And can we use all of them in our iframe tests (e.g when we access nav properties to detect lies etc)

see CB test page
- thanks document root
- iframe test 2 - with URL
- iframe test 3 - violating SOP
- iframe test 4 - access 1
- iframe test 5 - access 2
- iframe test 6 - access 3

It is noteworthy that Brave fails on tests 4/5/6 (and offscreen worker: edit this is fixed in Brave Nightly) - which makes me think that web ext APIs may be limited in these cases (to detect Web Ext lies), and indeed that Brave's randomizing of e.g. hardware concurrency leak in these iframes

Assigning to @abrahamjuliot 😬

**Comment by abrahamjuliot** [comment]
I'm reviewing [/test/test.js](https://canvasblocker.kkapsner.de/test/test.js) and [/test/test.html](https://canvasblocker.kkapsner.de/test/test.html).

Tests 1-3 call the method `iframeTest` which takes a `testNode` argument (a container that contains a child iframe). This can be reduced to 3 iframe tests.
```html
<!-- these can be loaded with the dom or injected via js -->
<iframe class="testnode" sandbox="allow-same-origin" style="display:none"></iframe> <!-- DocumentRoot -->
<iframe class="testnode" src="?" sandbox="allow-same-origin" style="display:none"></iframe> <!-- with URL -->
<iframe class="testnode" src="//example.org" style="display:none"></iframe> <!-- violating SOP -->
```
```js
const testNodes = document.querySelectorAll('.testnode')
const iframeContentWindows = [...testNodes].map(node => node.contentWindow)
iframeContentWindows.forEach(contentWindow => {
    // do something with contentWindow
})
```

Access tests 4-6 correspond to `dynamicIframeTest...` (1, 2 and 3). This can be reduced to 1 function.
```js
function getDynamicIframeWindow({ context, nestIframeInContainerDiv = false }) {
    const elementName = nestIframeInContainerDiv ? 'div' : 'iframe'
    const length = context.length
    const element = document.createElement(elementName)
    document.body.appendChild(element)
    if (nestIframeInContainerDiv) {
        element.innerHTML = '<iframe></iframe>'
    }
    const iframeWindow = context[length]
    document.body.removeChild(element)
    return iframeWindow
}

getDynamicIframeWindow({ context: frames })
getDynamicIframeWindow({ context: window })
getDynamicIframeWindow({ context: window, nestIframeInContainerDiv: true })
```

Together, these 6 tests can reduced to 1 function, without the use of `iframe.contentWindow`.

```js
function getDynamicIframeWindow({
    context,
    source = '',
    nestIframeInContainerDiv = false,
    violateSameOriginPolicy = true,
    display = false
}) {
    const elementName = nestIframeInContainerDiv ? 'div' : 'iframe'
    const length = context.length
    const element = document.createElement(elementName)
    document.body.appendChild(element)

    if (!display) {
        element.setAttribute('style', 'display:none')
    }

    if (nestIframeInContainerDiv) {
        const attributes = `
    	    ${source ? `src=${source}` : ''}
            ${violateSameOriginPolicy ? '' : `sandbox="allow-same-origin"`}
        `
        element.innerHTML = `<iframe ${attributes}></iframe>`
    } else if (!violateSameOriginPolicy) {
        element.setAttribute('sandbox', 'allow-same-origin')
        if (source) {
            element.setAttribute('src', source)
        }
    } else if (source) {
        element.setAttribute('src', source)
    }

    const iframeWindow = context[length]
    document.body.removeChild(element)
    return iframeWindow
}

getDynamicIframeWindow({
    context: window,
    violateSameOriginPolicy: false
}) // DocumentRoot

getDynamicIframeWindow({
    context: window,
    source: '?',
    violateSameOriginPolicy: false
}) // with URL

getDynamicIframeWindow({
    context: window,
    source: '//example.org'
}) // violating SOP

getDynamicIframeWindow({
    context: window
}) // access 1

getDynamicIframeWindow({
    context: frames
}) // access 2

getDynamicIframeWindow({
    context: window,
    nestIframeInContainerDiv: true
}) // access 3

```

**Comment by Thorin-Oakenpants** [comment]
Awesome. And props to @kkapsner

**Comment by Thorin-Oakenpants** [comment]
^ bump @abrahamjuliot :)

**Comment by abrahamjuliot** [comment]
Looks good. 

> SOP error

The SOP error requires waiting till the example site is loaded in the frame, and then to  trigger the error,  access `contentWindow.document` or `contentDocument`. In the current test, we remove the iframe before the example site is loaded. Here's a modified (a bit messy) [test](https://jsfiddle.net/5Luqko8b) that waits for the iframe to load. The error may be seen in the console. I chose the load event over set timeout because memory, cpu and network availability/capacity will require an unknown number of milliseconds to load. For example, through repeat refreshing, I get the SOP error on CB at random.

> Brave

I modified this [side test](https://jsfiddle.net/evLwnuxs) to check canvas and like the CB test, the main window does not match the access frames. It appears hardwareConcurrency in Brave is fine tuned in this area and the canvas remote frame leak is not caught on braves [test pages](https://dev-pages.bravesoftware.com/farbling.html).

**Comment by Thorin-Oakenpants** [comment]
OK, that's good - I thought I might have done a boo boo. I can drop the SOP test.

FYI: @pes10k emailed me to tell me that Brave have fixed the issue in Nighty (I told them). He told why these iframes didn't get "hooked" - but that doesn't explain why hardwareConcurrency was. The "farbling" would, _I assume_, be assigned and the seed remembered etc via a single point .. maybe @pes10k can explain .. something doesn't sound right (i.e canvas leaked but hardwareConcurrency didn't). Either our test is whack or there's something different about how Brave "detects" some items vs others

**Comment by kkapsner** [comment]
> Together, these 6 tests can be reduced to 1 function, without the use of `iframe.contentWindow`.

Nice find. This test page is one of my oldest and might need a workover. But testing contentWindow should still be done (at least for the CB test as it's the most common use of the iframe "bypass" - for TZP I'm not quite sure).

**Comment by pes10k** [comment]
Hi all, just wanted to follow up about what @Thorin-Oakenpants mentioned.

We have a fix, but it hasn't landed in nightly yet (that was my error in what i emailed @Thorin-Oakenpants). If you're interested, here are the details: https://github.com/brave/brave-core/pull/6941. The issue is that in blink space, we need to reach back up to the top level frame (to get the top level eTLD+1) in different ways.

This is all very deep in the weeds, but the short version of it is that we have a direct reference to the local document / frame in the hardwareConcurrency case, so figuring out the right seed to use is easy (see https://github.com/brave/brave-core/pull/6156/files).  In the canvas case, its more complicated, because we have only an indirect reference to it through the blink execution context (loosely, the blink abstraction for the document, service worker, web worker, etc), which requires us to get to the immediate frame through a different path when the frame is disconnected, which we hadn't considered before (see https://github.com/brave/brave-core/pull/6941/files#diff-3f73fe628fa3434c5a05ef66a4382f56b8af3630a0299fba65a4cf067ee83895R90).

TL;DR;, simplifying, handwave, etc, hardwareConcurrency worked because we (kinda) reimplement hardwareConcurrency, but have to work within the upstream implementation of the canvas methods.

Last, while do have a single system that handles "which per-session, per-etld+1 seed should i be using now" (see https://github.com/brave/brave-core/blob/master/chromium_src/third_party/blink/renderer/core/execution_context/execution_context.h), but we need to get to it in different ways, given existing blink abstractions.

Hope that helps explain what went wrong!

**Comment by Thorin-Oakenpants** [comment]
Thanks for the clarification and overview @pes10k

**Comment by Thorin-Oakenpants** [comment]
so what are better names for these iframes (I dropped violating SOP)?

access 1,2,3 don't mean much. The last one is nested, right?

![iframes](https://user-images.githubusercontent.com/16656956/100683449-4ad8ad80-3370-11eb-820c-48572906fd65.png)

**Comment by kkapsner** [comment]
What about "window access" and "frames access"?

**Comment by Thorin-Oakenpants** [comment]
so looking at the network inspector I see this
![violateSOP](https://user-images.githubusercontent.com/16656956/101270656-0afc3680-3773-11eb-8b42-47dccf749370.png)

generic:355 is this lot
```js
function getDynamicIframeWindow({
	context,
	source = "",
	test = "",
	nestIframeInContainerDiv = false,
	violateSameOriginPolicy = true,
	display = false
}) {
	const elementName = nestIframeInContainerDiv ? 'div' : 'iframe'
	const length = context.length
	const element = document.createElement(elementName)
	document.body.appendChild(element)
	if (!display) {
		element.setAttribute('style', 'display:none')
	}
	if (nestIframeInContainerDiv) {
	const attributes = `
		${source ? `src=${source}` : ''}
			${violateSameOriginPolicy ? '' : `sandbox="allow-same-origin"`}
		`
		element.innerHTML = `<iframe ${attributes}></iframe>`
	} else if (!violateSameOriginPolicy) {
		element.setAttribute('sandbox', 'allow-same-origin')
		if (source) {
			element.setAttribute('src', source) // <<<<< this is line 355
		}
	} else if (source) {
		element.setAttribute('src', source)
	}
	const iframeWindow = context[length]

	// do stuff

	document.body.removeChild(element)

}

```

I don't use violating SOP, and I don't see any scary red blocks with cb tests or at creep. Can I remove the offending code? Which bits (I'm afraid I might break the integrity of something)? why is `violateSameOriginPolicy` default true?

@abrahamjuliot

**Comment by Thorin-Oakenpants** [comment]
@abrahamjuliot (who is always busy) ... I also have a couple of questions

I put this in `generic.js` and pass a `test` parameter, so I can use it for other sections - but I want to check that if I call this multiple times, that each test won't interfer with the other: like referring to an iframe in `test="headers"` that was closed in `test="ua"`

I assume that each process can't start until the previous one finishes, or they each have their own iframe process ids or something?

Am I better off putting this function (renamed for each test) in each section js file

**Comment by abrahamjuliot** [comment]
If the function is called multiple times, the page may suffer performance issues due to `innerHTML` dom repaint/reflows. In any case, since the function scope holds a reference to the `element`, it won't clash with other executions of the same script. Ideally, the function should be called once and then the results can be passed on to the ua and other tests, however, that requires each test await the result. That's no biggy if the result is not a promise, but testing iframe's with a `source` will require awaiting the iframe to load. 

On that error, `source: "?"` triggers line 355 (regardless of `violateSameOriginPolicy`). I'm not sure, but it appears the error is thrown if the source is not loaded. CB loads the iframe source of test 1-3 within the html document. That approach is ideal for these 3 tests. I think a quick solution would be to remove test 1 and 2 and tests those in a seperate function. Testing an iframe with a source is useful since this context may not be protected. We could also wait for the iframes to load before accessing their windows.

This modified script (for example) will wait for the frames to load. 3 things added: 
- return a `promise`
- listen to `iframeWindow` `load` event and `resolve` there
- `contentWindow` window option for tests 1 and 2 (per [comment](https://github.com/arkenfox/TZP/issues/17#issuecomment-716067425))
```js
function getDynamicIframeWindow({
    context,
    source = "",
    test = "",
    contentWindow = false,
    nestIframeInContainerDiv = false,
    violateSameOriginPolicy = true,
    display = false
}) {
    return new Promise(resolve => {
        try {
            const elementName = nestIframeInContainerDiv ? 'div' : 'iframe'
            const length = context.length
            const element = document.createElement(elementName)
            document.body.appendChild(element)
            if (!display) {
                element.setAttribute('style', 'display:none')
            }
            if (nestIframeInContainerDiv) {
                const attributes = `
		${source ? `src=${source}` : ''}
			${violateSameOriginPolicy ? '' : `sandbox="allow-same-origin"`}
		`
                element.innerHTML = `<iframe ${attributes}></iframe>`
            } else if (!violateSameOriginPolicy) {
                element.setAttribute('sandbox', 'allow-same-origin')
                if (source) {
                    element.setAttribute('src', source)
                }
            } else if (source) {
                element.setAttribute('src', source)
            }


            const iframeWindow = contentWindow ? element.contentWindow : context[length]

            iframeWindow.addEventListener('load', () => {
                let res = []
                let navigator = iframeWindow.navigator

                if (test == "ua") {
                    let list = ['userAgent', 'appCodeName', 'appName', 'product', 'appVersion',
                            'oscpu', 'platform', 'buildID', 'productSub', 'vendor', 'vendorSub'
                        ],
                        r = ""

                    for (let i = 0; i < list.length; i++) {
                        try {
                            r = navigator[list[i]]
                        } catch (e) {
                            r = zB0
                        }
                        if (r == "") {
                            r = "empty string"
                        }
                        if (r == "undefined") {
                            r = "undefined string"
                        }
                        if (r == undefined) {
                            r = "undefined value"
                        }
                        res.push(list[i] + ":" + r)
                    }
                    res.sort()
                }
                document.body.removeChild(element)
                return resolve(res)
            })
        } catch (error) {
            // handle error
            console.error(error)
            return resolve()
        }
    })
}

Promise.all([
    getDynamicIframeWindow({
        context: window,
        contentWindow: true,
        violateSameOriginPolicy: false,
        test: "ua"
    }), // DocumentRoot
    getDynamicIframeWindow({
        context: window,
        contentWindow: true,
        source: "?",
        violateSameOriginPolicy: false,
        test: "ua"
    }), // with URL
    getDynamicIframeWindow({
        context: window,
        test: "ua"
    }), // window access
    getDynamicIframeWindow({
        context: frames,
        test: "ua"
    }), // iframe access
    getDynamicIframeWindow({
        context: window,
        nestIframeInContainerDiv: true,
        test: "ua"
    }), // nested
]).then(results => {
    return console.log(results)
})

```

**Comment by Thorin-Oakenpants** [comment]
that's super ultra slow: ms since I reset the global start timer for accumulative time

🔻 original

page load: about so about 20-30 ms from first request to finish
```
START iframe promises 181
ORIG iframe function promise 185
ORIG iframe function promise 190
ORIG iframe function promise 192
ORIG iframe function promise 195
ORIG iframe function promise 196
```

global rerun: 20-30
section rerun: about 30-40


🔻 the new one with listeners

page load: about 300 ms from first request to finish
```
START iframe promises 178
NEW iframe function promise 282
NEW iframe function promise 325
NEW iframe function promise 327
NEW iframe function promise 329
NEW iframe function promise 476
```

global rerun: between 400-700 <- At times it's almost the last thing to finish, even though it's one of the first things to be called
section rerun: about 180'ish (nothing else needs to run)

I'm not actually loading a source. I might do that (load a iframe source) for other iframe tests, probably behind a [click here] button, and I don't mind a different function for that

So now what. Leave it as it is, it's not actually breaking anything?

**Comment by abrahamjuliot** [comment]
Leave as is sounds good. You could add the `contentWindow: true` option for tests 1 and 2, and this line:
```js
const iframeWindow = contentWindow ? element.contentWindow : context[length]
```

**Comment by Thorin-Oakenpants** [comment]
done: https://github.com/arkenfox/TZP/commit/234a43ac622f6d1f93362b2cb56385c5a851a542

it till comes up with the same stack trace :-( 
![ugh](https://user-images.githubusercontent.com/16656956/101303044-17e15e80-3835-11eb-848f-bb736d7abeed.png)

**Comment by Thorin-Oakenpants** [comment]
PS: It only happens in FF79+

**Comment by Thorin-Oakenpants** [comment]
nah, that broke it (using chameleon)
![ugh](https://user-images.githubusercontent.com/16656956/101303708-896ddc80-3836-11eb-9f15-32a8a8ef5126.png)

**Comment by abrahamjuliot** [comment]
That's good, right? It shows Chameleon only protects `iframe.contentWindow`.

**Comment by Thorin-Oakenpants** [comment]
not AFAIK .. because it leaks without the changes ... so you tell me if it's checking the right contexts etc

**Comment by abrahamjuliot** [comment]
True, Chameleon leaks without the `contentWindow` test. If an extension protects the `iframe.window` context and not the `iframe.contentWindow` context, then the test could be useful. I'm not sure if that is a thing. 

Here are a few more extensions running with `contentWindow` tests. The nested test is a nice catch all. 

User-Agent Switcher
![image](https://user-images.githubusercontent.com/6946045/101308004-a9cd7500-37fd-11eb-9086-cb12702319fb.png)

Both Random User-Agent & User-Agent Switcher and Manager
![image](https://user-images.githubusercontent.com/6946045/101307872-55c29080-37fd-11eb-98ea-f739199e5a89.png)

Cydec
![image](https://user-images.githubusercontent.com/6946045/101307613-c3ba8800-37fc-11eb-8eba-25cbbc01762f.png)

Trace
![image](https://user-images.githubusercontent.com/6946045/101307734-00867f00-37fd-11eb-984e-832d7f6ee101.png)

**Comment by Thorin-Oakenpants** [comment]
So now I'm confused ... using contentWindow is protected by fluke or design - e.g. chameleon and trace? And in fact there are seven tests (eight if you count violating SOP) ? right?

What about spawning nested workers from nested iframes via iframe access?

**Comment by Thorin-Oakenpants** [comment]
@sereneblue Hi, do you do *any* protection of iframes for navigator user agent properties? thanks

**Comment by sereneblue** [comment]
@Thorin-Oakenpants Chameleon does protect `contentWindow` by design. I'm working on improving it to pass some of the other tests.

**Comment by Thorin-Oakenpants** [comment]
> does protect contentWindow by design

cool, that clears that up

> If an extension protects the `iframe.window` context and not the `iframe.contentWindow` context, then the test could be useful. I'm not sure if that is a thing.

I'm not sure if it's worth expanding the iframe tests to have both tests 1 + 2 with contentWindow as true and false. What do you think abraham - it wouldn't only add about 5 ms to the ua test (I don't plan on doing all iframe methods elsewhere)

**Comment by Thorin-Oakenpants** [comment]
nvm .. I expanded it to seven tests

**Comment by abrahamjuliot** [comment]
Looks good.

> spawning nested workers from nested iframes via iframe access

Creating the worker via the nested iframe access can bypass attempts to block or tamper with the worker prototype, but then the worker scope is the same with or without the iframe, and as far as I know, the nested iframe will not bypass `Content-Security-Policy` where workers are blocked.
